### DISCUSSION TITLE
App Router, SPA mode & Export

### DISCUSSION THREAD

**Discussion by ytsruh** [vercel/next.js#49517] [App Router]
Hi all,

I have looked around for an answer on this but to no avail. I'm currently running a SPA in the App router (eg all client components) & using Next Export. Everything works as expected except for dynamic routes.

On a dynamic route if I directly navigate to it then I get a 404 error in the build. Eg if I type in /post/1 then the 404 is returned. When running in dev mode this is fine as I assume there is some route interception being done, but how to do this in a production environment? I assume it needs some rewrites but it's not clear which file in the export it should be re-written to.

Thanks

**Comment by icyJoseph** [comment]
Hi,

This has always been an "issue", even in pages dir. At least, with the information given.

So your issue is that once you navigate to `/some/route`, on your exported app, it yields 404, right? However, navigating from `/` to `/some/route`, works. And reloading the page on `/some/route`, yields 404 too.

This is because your static assets server is failing to point `/some/route` to `/some/route.html` which is what the export process outputs. 

On this discussions several people have shared their server configurations: https://github.com/vercel/next.js/discussions/10522

For example, if I have an app with `/` and `/blog` routes:

```console
out
├── 404.html
├── _next
├── blog.html
├── blog.txt
├── favicon.ico
├── index.html
├── index.txt
├── next.svg
└── vercel.svg
```

This is what the output folder has. Notice there's `blog.html`, so my static server would need to serve `blog.html`, when `/blog` is requested.

For example if I use: https://github.com/vercel/serve-handler as my static assets server, it'll handle `/blog` properly.

So you need to figure out how to tell your static assets server to do that. 

One additional, generic, caveat, if you had a single entry point app before, your server might already have some configuration to send, all not found requests to `index.html`, which is not what we'd want in a multiple entry point SPA such as Next.js exported.

**Reply by ytsruh** [reply]
Thanks for the excellent explanation. My hope was that this had been 'fixed' with App Router given it is less Next & more plain React.

This is a bit of a crazy situation then because NextJS 'SPA mode' doesn't really support dynamic routing at all - yet the docs suggest that it does. I have Dan Abramov telling me to use a Framework (he actually told me this on Twitter) yet this very basic use case is not supported.

My issue is different to the one you linked to. I'm having issues rendering dynamic routes that are not known at build time. Eg im pulling blog posts out of a CMS for instance - eg /blog/some-title-from-cms - the issue is that no file is generated from the Next Export and therefore it appears to not be possible to redirect a user server side should they navigate to this page

**Reply by icyJoseph** [reply]
Aha, but you are talking instead about having paths such as `/some/[slug]` - correct?

Doesn't this apply to your case? https://nextjs.org/docs/app/building-your-application/deploying/static-exports#dynamic-fetching-with-client-components

**Reply by ytsruh** [reply]
Yes thats what I'm referring to (apologies should have linked that). The trouble with this is that there is no page created for that slug - see screenshot.

<img width="269" alt="Screenshot 2023-05-09 at 17 00 52" src="https://github.com/vercel/next.js/assets/46472228/d45c76ee-d10b-4f3c-a350-6773c8256564">

So if a user types /post/a-valid-slug the app returns a 404

**Reply by icyJoseph** [reply]
That's weird, what version of Next.js are you using?

Mine crashes out 😅 


```console
> Build error occurred
Error: invariant: Undefined export worker for app dir, this is a bug in Next.js.
```

**Reply by ytsruh** [reply]
Oh hahaah! You need the canary version. See this [issue](https://github.com/vercel/next.js/issues/49059#issuecomment-1539074712) (another pain point I had)

**Reply by icyJoseph** [reply]
Yah, got it, but as you say, there's no `/blog/[slug]` - I'd report this as a bug though. Or at least it deservers an explanation because it deviates from that the docs say.

**Reply by ytsruh** [reply]
Honestly....thank you! I thought I was going crazy. I'll report as a bug now

**Reply by icyJoseph** [reply]
Try to prepare a small reproduction repo. It might be harder to do it with Stackblitz, so a GitHub repo is probably better.

**Reply by ytsruh** [reply]
Just done it - https://github.com/vercel/next.js/issues/49538 - with a reproduction repo. Thanks for your help on this, really appreciate it
